rules/Naming/PropertyRenamer/PropertyPromotionRenamer.php

```diff
@@ -0,0 +1,118 @@
+    ) {
+    }
+
+    public function renamePropertyPromotion(ClassLike $classLike): void
+    {
+        if (! $this->phpVersionProvider->isAtLeastPhpVersion(PhpVersionFeature::PROPERTY_PROMOTION)) {
+            return;
+        }
+
+        $constructClassMethod = $classLike->getMethod(MethodName::CONSTRUCT);
+        if (! $constructClassMethod instanceof ClassMethod) {
+            return;
+        }
+
+        foreach ($constructClassMethod->params as $param) {
+            if ($param->flags === 0) {
+                continue;
+            }
+
+            // promoted property
+            $desiredPropertyName = $this->matchParamTypeExpectedNameResolver->resolve($param);
+            if ($desiredPropertyName === null) {
+                continue;
+            }
+
+            $currentParamName = $this->nodeNameResolver->getName($param);
+            if ($this->isNameSuffixed($currentParamName, $desiredPropertyName)) {
+                continue;
+            }
+
+            $this->renameParamVarName($classLike, $constructClassMethod, $desiredPropertyName, $param);
+        }
+    }
+
+    private function renameParamVarName(
+        ClassLike $classLike,
+        ClassMethod $classMethod,
+        string $desiredPropertyName,
+        Param $param
+    ): void {
+        $classMethodPhpDocInfo = $this->phpDocInfoFactory->createFromNodeOrEmpty($classMethod);
+
+        $currentParamName = $this->nodeNameResolver->getName($param);
+
+        $this->propertyFetchRenamer->renamePropertyFetchesInClass(
+            $classLike,
+            $currentParamName,
+            $desiredPropertyName
+        );
+
+        /** @var string $paramVarName */
+        $paramVarName = $param->var->name;
+        $this->renameParamDoc($classMethodPhpDocInfo, $param, $paramVarName, $desiredPropertyName);
+        $param->var->name = $desiredPropertyName;
+    }
+
+    private function renameParamDoc(
+        PhpDocInfo $phpDocInfo,
+        Param $param,
+        string $paramVarName,
+        string $desiredPropertyName
+    ): void {
+        $paramTagValueNode = $phpDocInfo->getParamTagValueNodeByName($paramVarName);
+        if (! $paramTagValueNode instanceof ParamTagValueNode) {
+            return;
+        }
+
+        $paramRename = $this->paramRenameFactory->createFromResolvedExpectedName($param, $desiredPropertyName);
+        if (! $paramRename instanceof ParamRename) {
+            return;
+        }
+
+        $this->paramRenamer->rename($paramRename);
+    }
+
+    /**
+     * Sometimes the bare type is not enough.
+     * This allows prefixing type in variable names, e.g. "Type $firstType"
+     */
+    private function isNameSuffixed(string $currentParamName, string $desiredPropertyName): bool
+    {
+        $currentNameLowercased = strtolower($currentParamName);
+        $expectedNameLowercased = strtolower($desiredPropertyName);
+
+        return str_ends_with($currentNameLowercased, $expectedNameLowercased);
+    }
```

Comment:
@samsonasik The renaming code in Rector is quite complicated, I could not find the logic I ferred to.

This is the logic I talked about:

- variable: `$superFood`, type: `Food` :heavy_check_mark: 
- variable: `$heatlhyFood`, type: `Food` :heavy_check_mark: 
- variable: `$superCat`, type: `Food` :x: 
- variable: `$superFood`, type: `SuperFood` :heavy_check_mark: